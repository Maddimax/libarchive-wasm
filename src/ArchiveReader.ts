/* eslint-disable no-underscore-dangle */
import type { LibarchiveWasm } from './libarchiveWasm';

export class ArchiveReader {
  public libarchive: LibarchiveWasm;

  public archive: number;

  public pointer: number;

  static FileTypes = {
    [`${0o170000}`]: 'Mount',
    [`${0o100000}`]: 'File',
    [`${0o120000}`]: 'SymbolicLink',
    [`${0o140000}`]: 'Socket',
    [`${0o020000}`]: 'CharacterDevice',
    [`${0o060000}`]: 'BlockDevice',
    [`${0o040000}`]: 'Directory',
    [`${0o010000}`]: 'NamedPipe',
  };

  constructor(libarchive: LibarchiveWasm, data: Int8Array, passphrase?: string) {
    const ptr = libarchive.module._malloc(data.length);
    libarchive.module.HEAP8.set(data, ptr);

    this.libarchive = libarchive;
    this.archive = libarchive.read_new_memory(ptr, data.length, passphrase as string);
    this.pointer = ptr;
  }

  free(): void {
    this.libarchive.read_free(this.archive);
    this.libarchive.module._free(this.pointer);

    this.libarchive = null as unknown as LibarchiveWasm;
    this.archive = null as unknown as number;
    this.pointer = null as unknown as number;
  }

  hasEncryptedData(): boolean | null {
    const code = this.libarchive.read_has_encrypted_entries(this.archive);
    return code < 0 ? null : !!code;
  }

  readData(size: number): Int8Array {
    const eptr = this.libarchive.module._malloc(size);
    const esize = this.libarchive.read_data(this.archive, eptr, size);
    const data = this.libarchive.module.HEAP8.slice(eptr, eptr + esize);
    this.libarchive.module._free(eptr);
    return data;
  }

  skipData(): void {
    this.libarchive.read_data_skip(this.archive);
  }

  nextEntryPointer(): number {
    return this.libarchive.read_next_entry(this.archive);
  }

  getEntryFiletype(ptr: number): string {
    return ArchiveReader.FileTypes[`${this.libarchive.entry_filetype(ptr)}`] || 'Invalid';
  }

  getEntryPathname(ptr: number): string {
    return this.libarchive.entry_pathname(ptr);
  }

  getEntrySize(ptr: number): number {
    return this.libarchive.entry_size(ptr);
  }

  isEntryEncrypted(ptr: number): boolean {
    return !!this.libarchive.entry_is_encrypted(ptr);
  }
}
