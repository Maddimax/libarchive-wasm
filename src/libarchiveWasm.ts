import libarchive from './libarchive';
import { wrapLibarchiveWasm } from './wrapLibarchiveWasm';

export type LibarchiveWasm = ReturnType<typeof wrapLibarchiveWasm>;

export function libarchiveWasm(...args: Parameters<typeof libarchive>): Promise<LibarchiveWasm> {
  return new Promise((resolve) => {
    libarchive(...args).then((mod) => {
      resolve(wrapLibarchiveWasm(mod));
    });
  });
}
