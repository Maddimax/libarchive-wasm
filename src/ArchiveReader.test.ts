import { readFile } from 'fs/promises';

import { libarchiveWasm } from './libarchiveWasm';
import { ArchiveReader } from './ArchiveReader';

const toEntries = (a: ArchiveReader): Record<string, unknown>[] => {
  const entries = [] as Record<string, unknown>[];
  for (;;) {
    const entryPointer = a.nextEntryPointer();
    if (entryPointer === 0) break;
    const pathname = a.getEntryPathname(entryPointer);
    const size = a.getEntrySize(entryPointer);
    const data = /\.md/.test(pathname) ? a.readData(size) : a.skipData();
    entries.push({
      filetype: a.getEntryFiletype(entryPointer),
      pathname,
      size,
      data: new TextDecoder().decode(data || undefined),
      encrypted: a.isEntryEncrypted(entryPointer),
    });
  }
  return entries;
};

describe('ArchiveReader', () => {
  it('test.zip', async () => {
    const data = await readFile('./archives/example.zip');
    const mod = await libarchiveWasm();
    const a = new ArchiveReader(mod, new Int8Array(data));
    expect(a.hasEncryptedData()).toBe(null);
    const entries = toEntries(a);
    expect(entries).toMatchSnapshot();
    expect(a.hasEncryptedData()).toBe(false);
    a.free();
  });

  it('encrypted.zip', async () => {
    const data = await readFile('./archives/encrypted.zip');
    const mod = await libarchiveWasm();
    const a = new ArchiveReader(mod, new Int8Array(data), 'Passw0rd!');
    expect(a.hasEncryptedData()).toBe(null);
    const entries = toEntries(a);
    expect(entries).toMatchSnapshot();
    expect(a.hasEncryptedData()).toBe(true);
    a.free();
  });
});
