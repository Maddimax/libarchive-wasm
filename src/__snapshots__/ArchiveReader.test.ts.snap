// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchiveReader encrypted.zip 1`] = `
Array [
  Object {
    "data": "",
    "encrypted": false,
    "filetype": "Directory",
    "pathname": "archive/",
    "size": 0,
  },
  Object {
    "data": "# archive

\`\`\`
archives/archive
archives/archive/README.md
archives/archive/dir
archives/archive/dir/symlink
archives/archive/dir/image.png
\`\`\`
",
    "encrypted": true,
    "filetype": "File",
    "pathname": "archive/README.md",
    "size": 144,
  },
  Object {
    "data": "",
    "encrypted": false,
    "filetype": "Directory",
    "pathname": "archive/dir/",
    "size": 0,
  },
  Object {
    "data": "",
    "encrypted": true,
    "filetype": "File",
    "pathname": "archive/dir/symlink",
    "size": 144,
  },
  Object {
    "data": "",
    "encrypted": true,
    "filetype": "File",
    "pathname": "archive/dir/image.png",
    "size": 1559,
  },
]
`;

exports[`ArchiveReader test.zip 1`] = `
Array [
  Object {
    "data": "",
    "encrypted": false,
    "filetype": "Directory",
    "pathname": "archive/",
    "size": 0,
  },
  Object {
    "data": "# archive

\`\`\`
archives/archive
archives/archive/README.md
archives/archive/dir
archives/archive/dir/symlink
archives/archive/dir/image.png
\`\`\`
",
    "encrypted": false,
    "filetype": "File",
    "pathname": "archive/README.md",
    "size": 144,
  },
  Object {
    "data": "",
    "encrypted": false,
    "filetype": "Directory",
    "pathname": "archive/dir/",
    "size": 0,
  },
  Object {
    "data": "",
    "encrypted": false,
    "filetype": "File",
    "pathname": "archive/dir/symlink",
    "size": 144,
  },
  Object {
    "data": "",
    "encrypted": false,
    "filetype": "File",
    "pathname": "archive/dir/image.png",
    "size": 1559,
  },
]
`;
